File: src/main/java/co/bitshifted/appforge/ignite/ui/UIRegistry.java
/*
 *
 *  * Copyright (c) 2022  Bitshift D.O.O (http://bitshifted.co)
 *  *
 *  * This Source Code Form is subject to the terms of the Mozilla Public
 *  * License, v. 2.0. If a copy of the MPL was not distributed with this
 *  * file, You can obtain one at http://mozilla.org/MPL/2.0/.
 *
 */

package co.bitshifted.appforge.ignite.ui;

import co.bitshifted.appforge.ignite.IgniteAppConstants;
import co.bitshifted.appforge.ignite.ctrl.*;
import javafx.fxml.FXMLLoader;
import javafx.scene.Parent;
import javafx.scene.control.MenuBar;
import javafx.scene.layout.BorderPane;
import javafx.stage.Stage;

import java.io.IOException;
import java.util.HashMap;
import java.util.Map;
import java.util.ResourceBundle;

public class UIRegistry {

    private static final UIRegistry INSTANCE;

    // Component names
    public static final String MAIN_PAGE = "main-page";
    public static final String MAIN_MENU = "main-menu";
    public static final String DEPLOYMENT_INFO_DLG = "deployment-info-dlg";
    public static final String DEPLOYMENT_INFO = "deployment-info";
    public static final String  APPLICATION_INFO_UI = "app-info";
    public static final String  APPLICATION_INFO_LINUX_UI = "app-info-linux";
    public static final String  APPLICATION_INFO_WINDOWS_UI = "app-info-windows";
    public static final String  APPLICATION_INFO_MAC_UI = "app-info-mac";
    public static final String SERVER_MANAGEMENT = "server-management";
    public static final String ADD_SERVER_PANE = "add-server-pane";
    public static final String PROJECT_BUTTONS_BAR = "project-button-bar";
    public static final String JVM_INFO = "jvm-info";
    public static final String RESOURCES_DATA = "resources-data";

    static {
        INSTANCE = new UIRegistry();
    }

    private final Map<String, Parent> componentMap;
    private Stage mainWindow;

    private UIRegistry() {
        componentMap = new HashMap<>();
    }

    public static UIRegistry instance() {
        return INSTANCE;
    }

    public void registerComponents() throws IOException {
        var bundle = ResourceBundle.getBundle(IgniteAppConstants.MESSAGE_BUNDLE_NAME);

        componentMap.put(PROJECT_BUTTONS_BAR, FXMLLoader.load(getClass().getResource("/fxml/project-button-bar.fxml"), bundle));
        loadWithController("/fxml/server-management.fxml", ServerManagementController.class, SERVER_MANAGEMENT, bundle);
        loadWithController("/fxml/new-server.fxml", AddServerController.class, ADD_SERVER_PANE, bundle);
        loadWithController("/fxml/app-info.fxml", AppInfoController.class, APPLICATION_INFO_UI, bundle);
        loadWithController("/fxml/app-info-linux.fxml", AppInfoLinuxController.class, APPLICATION_INFO_LINUX_UI, bundle);
        loadWithController("/fxml/app-info-windows.fxml", AppInfoWindowsController.class, APPLICATION_INFO_WINDOWS_UI, bundle);
        loadWithController("/fxml/app-info-mac.fxml", AppInfoMacController.class, APPLICATION_INFO_MAC_UI, bundle);
        loadWithController("/fxml/deployment-info-dlg.fxml", DeploymentInfoDlgController.class, DEPLOYMENT_INFO_DLG, bundle);
        loadWithController("/fxml/deployment-info.fxml", DeploymentInfoController.class, DEPLOYMENT_INFO, bundle);
        loadWithController("/fxml/jvm.fxml", JvmInfoController.class, JVM_INFO, bundle);
        loadWithController("/fxml/resources.fxml", ResourcesController.class, RESOURCES_DATA, bundle);
        var mainMenu = (MenuBar)FXMLLoader.load(getClass().getResource("/fxml/main-menu.fxml"), bundle);
        componentMap.put(MAIN_MENU, mainMenu);
        // load main window last, to make sure all children are loaded
        var mainPage = (BorderPane)FXMLLoader.load(getClass().getResource("/fxml/main.fxml"), bundle);
        componentMap.put(MAIN_PAGE, mainPage);
        // layout
        mainPage.setTop(mainMenu);
    }

    public Parent createView(String fxmlFile, ResourceBundle bundle, Object controller) throws IOException {
        var loader = new FXMLLoader(getClass().getResource(fxmlFile));
        loader.setResources(bundle);
        loader.setController(controller);
        return loader.load();
    }

    public Parent getComponent(String name) {
        return componentMap.get(name);
    }

    private void loadWithController(String fxmlFile, Class clazz  , String key, ResourceBundle bundle) throws IOException {
        var loader = new FXMLLoader(getClass().getResource(fxmlFile));
        loader.setResources(bundle);
        loader.setController(ControllerRegistry.instance().getController(clazz));

        componentMap.put(key, loader.load());
    }

    public void setMainWindow(Stage mainWindow) {
        this.mainWindow = mainWindow;
    }

    public Stage getMainWindow() {
        return mainWindow;
    }
}

File: src/main/kotlin/co/bitshifted/xapps/ignite/ui/UIRegistry.kt
/*
 * Copyright (c) 2020. Bitshift (http://bitshifted.co)
 *
 * This Source Code Form is subject to the terms of the Mozilla Public
 * License, v. 2.0. If a copy of the MPL was not distributed with this
 * file, You can obtain one at http://mozilla.org/MPL/2.0/.
 */

package co.bitshifted.xapps.ignite.ui

import co.bitshifted.xapps.ignite.ctrl.AddServerController
import co.bitshifted.xapps.ignite.ctrl.ControllerRegistry
import co.bitshifted.xapps.ignite.ctrl.NewProjectDialogController
import co.bitshifted.xapps.ignite.ctrl.ProjectInfoController
import javafx.fxml.FXMLLoader
import javafx.scene.Parent
import javafx.stage.Stage
import java.util.*

object UIRegistry {

    const val MAIN_PAGE = "main_page"
    const val START_PANE = "start_pane"
    const val PROJECT_INFO_PANE = "project_info_pane"
    const val PROJECT_BUTTON_BAR = "project_button_bar"
    const val APP_INFO_PANE = "app_info_pane"
    const val DEPENDENCY_INFO_PANE = "dependency_info_pane"
//    const val PLATFORM_DEPENDENCY_PANE = "platform_dependency_pane"
//    const val ADD_MAVEN_DEPENDENCY_PANE = "add_maven_dependency_pane"
    const val ADD_SERVER_PANE = "add_server_pane"
    const val JVM_PROPERTIES_PANE = "jvm_properties_pane"


    private val componentMap = mutableMapOf<String, Parent>()
    private lateinit var mainWindow : Stage

    fun loadComponents() {
        val bundle = ResourceBundle.getBundle("i18n/strings")
        componentMap[START_PANE] = FXMLLoader.load(javaClass.getResource("/fxml/start-pane.fxml"), bundle)
        loadWithController("/fxml/project-info.fxml", ProjectInfoController::class.java, PROJECT_INFO_PANE, bundle)
        componentMap[PROJECT_BUTTON_BAR] = FXMLLoader.load(javaClass.getResource("/fxml/project-button-bar.fxml"))
        componentMap[APP_INFO_PANE] = FXMLLoader.load(javaClass.getResource("/fxml/app-info.fxml"))
        componentMap[DEPENDENCY_INFO_PANE] = FXMLLoader.load(javaClass.getResource("/fxml/dependency-info.fxml"))
//        componentMap[PLATFORM_DEPENDENCY_PANE] = FXMLLoader.load(javaClass.getResource("/fxml/platform-deps-info.fxml"))
        componentMap[JVM_PROPERTIES_PANE] = FXMLLoader.load(javaClass.getResource("/fxml/jvm-properties.fxml"))
//        loadWithController("/fxml/maven-dependency-pane.fxml", NewMavenDependencyController::class.java, ADD_MAVEN_DEPENDENCY_PANE, bundle)
        loadWithController("/fxml/new-server.fxml", AddServerController::class.java, ADD_SERVER_PANE, bundle)

        // load main window last, to make sure all children are loaded
        componentMap[MAIN_PAGE] = FXMLLoader.load<Parent>(javaClass.getResource("/fxml/main.fxml"))
    }

    /**
     * Used to load FXML file with specified resource bundle.
     *
     */
    private fun <T> loadWithController(fxmlFile : String, clazz : Class<T>, key : String, bundle : ResourceBundle) {
        val loader = FXMLLoader(javaClass.getResource(fxmlFile))
        loader.resources = bundle
        loader.setController(ControllerRegistry.getController(clazz))

        componentMap[key] = loader.load()
    }

    fun getComponent(name : String) : Parent? = componentMap[name]

    fun setMainWindow(stage : Stage) { mainWindow = stage }
    fun getMainWindow() : Stage =   mainWindow
}